### DISCUSSION TITLE
Generate the JS to invoke the custom action/api

### DISCUSSION THREAD

**Discussion by paulhasell** [rappen/CustomActionTester#25] [Ideas]
Hi, it would be useful if there was an option to generate the JS to invoke the action/api as I often write actions to respond to UX events for realtime data/calculations.

There are 2 styles for invoking from JS; using Xrm.WebApi or by posting to the WebApi URL. Ideally both would be handy though I prefer the later as it's fewer lines of code (I use **fetch** mostly now).

The main gain would be around how to pass the different parameter types in each style as documentation seems limited/confusing on the structures/standards...though that may just be me :)

**Comment by rappen** [comment]
Hi @paulhasell - interesting!

Currently, my head is a bit off this area, but I would like to add a few more thoughts from @MarkMpn, @BenediktBergmann, @daryllabar, and maybe even @MattB-msft as I know he has some good ideas on preference use :)

**Comment by drivardxrm** [comment]
@paulhasell (cc. @rappen )
I think that the [Dataverse REST Builder](https://github.com/GuidoPreite/DRB) tool from @GuidoPreite does exactly what tou are asking (and a lot more).
The tool is available standalone or in the XrmToolBox

Just choose the Execute Custom API (or Action) Request type and follow along to generate the JS boilerplate

![image](https://user-images.githubusercontent.com/38399134/148688166-dbf0b57b-5638-4c9d-9ee7-5b0ad4da62e8.png)

PS : @rappen @GuidoPreite It would be great if we could integrate our 3 tools together in the toolbox : 
Custom API Manager + Custom API Tester + Dataverse REST Builder ;)

**Reply by GuidoPreite** [reply]
thanks @drivardxrm for mentioning DRB. One thing to consider here are the parameter types, if the type is Entity or EntityCollection, with DRB is possible only to pass the primary id (similar as the EntityReference, inside DRB this part uses the same terminology from the metadata, mscrm.crmbaseentity) but it possible to pass additional columns (and if it is expando they can also be not connected to a specific table).
So if the Custom API/Custom Action uses the standard parameters the code generated by DRB works, if the types I mentioned above are involved then the code must be manually adapted to include the additional columns

**Comment by paulhasell** [comment]
Hi All,

I couldn't get Dataverse REST Builder to load, it blew up trying to access a WebView2 control...I'll try again later to see if it's just me.

I ended up going a 'simplified' route for calling the API, direct JSON:

        private async alignToServiceCalendar(dueDate?: Date) {
            var context = this.props.xrm.Utility.getGlobalContext();
            var clientUrl = context.getClientUrl();
            var version = context.getVersion();
            var api = `${clientUrl}/api/data/v${version}/nfp_aligntoservicecalendar`;
            var payload = JSON.stringify({
                PaymentProcessor: { ***@***.***': 'Microsoft.Dynamics.CRM.nfp_paymentprocessor', 'nfp_paymentprocessorid': this.state.paymentProcessorId },
                Date: dueDate.toISOString()
            });
            console.log(api);
            console.log(payload);
            var response = await fetch(api, {
                method: 'post',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: payload
            }).then((response) => {
                console.log(response);
                return response.json();
            });
            var alignedDate = new Date(response.AlignedDate);
            this.collectionDateAttr.setValue(alignedDate);
            this.setState({
                collectionDate: alignedDate,
                calculating: false
            })
        }
​This is a lower code option than the examples I've seen to date using XMLHTTPRequest...though I may not have looked wide enough 🙂


Many Regards




________________________________


PAUL HASELL
DYNAMICS 365 PRODUCT DEVELOPMENT LEAD
m-hance
Tel: 0044 161 406 2335
Mobile: 0044 7917 899 007
Email: ***@***.***
www.m-hance.com



Disclaimer: To read our email disclaimer please visit our website.
From: Guido Preite ***@***.***>
Sent: 09 January 2022 15:25
To: rappen/CustomActionTester ***@***.***>
Cc: Paul Hasell ***@***.***>; Mention ***@***.***>
Subject: [EXT]: Re: [rappen/CustomActionTester] Generate the JS to invoke the custom action/api (Discussion #25)


Caution: External email

This e-mail originated outside m-hance, unless you recognise the sender and know the content is safe, do not click links or open attachments.



thanks @drivardxrm<https://github.com/drivardxrm> for the mentioning DRB. One thing to consider here are the parameter types, if the type is Entity or EntityCollection, with DRB is possible only to pass the primary id (similar as the EntityReference, inside DRB this part uses the same terminology from the metadata, mscrm.crmbaseentity) but it possible to pass additional columns (and if it is expando they can also be not connected to a specific table).
So if the Custom API/Custom Action uses the standard parameters the code generated by DRB works, if the types I mentioned above are involved then the code must be manually adapted to include the additional columns

—
Reply to this email directly, view it on GitHub<https://github.com/rappen/CustomActionTester/discussions/25#discussioncomment-1935482>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE3ZBFUAHXRYWM4I7SKI7ADUVGSILANCNFSM5KUUXIKQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by GuidoPreite** [reply]
@paulhasell for the XrmToolBox version make sure you update XrmToolBox as Dataverse REST Builder requires the most recent version (1.2021.12.53), if you still have issues loading it with that version please open an issue inside DRB repository with screenshots and some details (like the exact Windows version you are running), I will try to replicate the issue on my side.
Keep in mind that Dataverse REST Builder is available also as managed solution, so you can also install it inside the instance containing the Custom API / Custom Action you want to invoke.
